algorithms/kernel/dbscan/oneapi/dbscan_dense_default_batch_ucapi_impl.i

```diff
@@ -0,0 +1,511 @@
+            {
+                continue;
+            }
+            coreIndices.get()[pos] = i;
+            pos++;
+        }
+    }
+
+    if (resultsToCompute & computeCoreObservations)
+    {
+        DAAL_CHECK_STATUS_VAR(ntCoreObservations->resize(nCoreObservations));
+        BlockDescriptor<algorithmFPType> coreObservationsRows;
+        DAAL_CHECK_STATUS_VAR(ntCoreObservations->getBlockOfRows(0, nCoreObservations, writeOnly, coreObservationsRows));
+        auto coreObservations = coreObservationsRows.getBuffer();
+
+        uint32_t pos = 0;
+        int result   = 0;
+        for (uint32_t i = 0; i < nRows; i++)
+        {
+            if (!isCore[i])
+            {
+                continue;
+            }
+            BlockDescriptor<algorithmFPType> dataRows;
+            DAAL_CHECK_STATUS_VAR(ntData->getBlockOfRows(i, 1, readOnly, dataRows));
+            auto data = dataRows.getBuffer();
+
+            Status st;
+            context.copy(UniversalBuffer(coreObservations), pos * nFeatures, UniversalBuffer(data), 0, nFeatures, &st);
+            DAAL_CHECK_STATUS_VAR(st);
+            pos++;
+            DAAL_CHECK_STATUS_VAR(ntData->releaseBlockOfRows(dataRows));
+        }
+        if (result)
+        {
+            return Status(services::ErrorMemoryCopyFailedInternal);
+        }
+    }
+
+    return Status();
+}
+
+template <typename algorithmFPType>
+Status DBSCANBatchKernelUCAPI<algorithmFPType>::compute(const NumericTable * x, const NumericTable * ntWeights, NumericTable * ntAssignments,
+                                                        NumericTable * ntNClusters, NumericTable * ntCoreIndices, NumericTable * ntCoreObservations,
+                                                        const Parameter * par)
+{
+    Status s;
+    auto & context        = Environment::getInstance()->getDefaultExecutionContext();
+    auto & kernel_factory = context.getClKernelFactory();
+
+    const algorithmFPType epsilon         = par->epsilon;
+    const algorithmFPType minObservations = par->minObservations;
+    const uint32_t minkowskiPower         = (uint32_t)2.0;
+    algorithmFPType epsP                  = 1.0;
+    for (uint32_t i = 0; i < minkowskiPower; i++) epsP *= epsilon;
+
+    NumericTable * ntData = const_cast<NumericTable *>(x);
+    if (ntData->getNumberOfRows() > static_cast<size_t>(UINT_MAX) || ntData->getNumberOfColumns() > static_cast<size_t>(UINT_MAX))
+    {
+    }
+    const uint32_t nRows = ntData->getNumberOfRows();
+    const uint32_t dim   = ntData->getNumberOfColumns();
+
+    BlockDescriptor<algorithmFPType> dataRows;
+    ntData->getBlockOfRows(0, nRows, readOnly, dataRows);
+    auto data = dataRows.getBuffer();
+
+    BlockDescriptor<int> assignRows;
+    DAAL_CHECK_STATUS_VAR(ntAssignments->getBlockOfRows(0, nRows, writeOnly, assignRows));
+    auto assignments = daal::oneapi::internal::UniversalBuffer(assignRows.getBuffer());
+    context.fill(assignments, undefined, &s);
```

Comment:
You don't have to explicitly convert `services::Buffer` to `UniversalBuffer`, its constructor is not explicit.
 ```cpp
auto assignments = assignRows.getBuffer();
context.fill(assignments, undefined, &s);
```

Reply:
Where the `undefined` is came from?

Reply:
Was discussed offline